### ISSUE TITLE
Opus files on the device are ignored

### ISSUE THREAD

**Issue by Daykm** [timusus/Shuttle#2]
Shuttle Version : Dev and store
Device: Pixel XL 7.1.1
Description: The app doesn't show any .opus file (http://opus-codec.org/). These files are in the ogg format but have a different extension. As far as I know, the Android platform has supported the codec since 5.0.
Steps: Add any .opus file to the devices Music directory. Open app.
Expected: The opus file would be an item in any list
Outcome: The file is not in any list

I'm interested in contributing so I'll start looking into what needs to be added for this but I'm making this issue to keep track of this feature.

**Comment by damongant** [comment]
I use `x-opus+ogg` files for my library. Rhythmbox sync works around this issue by naming all copied files `.ogg`, but they're not reencoded so treating them as `ogg` should work fine.

**Comment by damongant** [comment]
Since the PR doesn't seem to work as intended, here might be some more info:

https://code.google.com/p/android/issues/detail?id=80729

**Comment by Daykm** [comment]
I think this might have to be part of the plans @timusus makes for weaning off MediaStore as he mentioned in the readme.

**Comment by timusus** [comment]
>As far as I know, the Android platform has supported the codec since 5.0.

No, Opus files are not yet supported. Google claimed to add support in Lollipop. They made some very minor changes, but AFAIK, they never actually added support - the MediaScanner doesn't add Opus to the MediaStore and the MediaPlayer doesn't play Opus.

**Comment by bobobo1618** [comment]
I actually managed to get Opus working.

Device: Google Pixel, running Android 7.1.1

Files: The result of running `ffmpeg -i <input> -c:a libopus -b:a 192k -vn -y <output>.opus.ogg`

I found however that ffmpeg, which I suspect most people are using, by default transcodes the album art as a Theora video stream. When album art is present, the file fails to play, when the album art is omitted (the `-vn` argument), the album is added to MediaStore and Shuttle is able to read its metadata and play it (and Shuttle also somehow gets album art?).

**Comment by timusus** [comment]
While I find this interesting, none of it is relevant to Shuttle's ability to play Opus files. Either Google actually add support for Opus via the `MediaScanner`, `MediaStore` and `MediaPlayer`, or Shuttle moves to a 3rd party audio player.

**Comment by bobobo1618** [comment]
What I'm saying is that Shuttle _is currently able to play Opus files_. Google _has_ added support for Opus via the MediaScanner, MediaStore and MediaPlayer.

I currently have ~3000 Opus files on my phone, running stock Android. Shuttle has added them all to its internal library and I'm able to view and play them inside Shuttle, as if they were any other format.

**Comment by bobobo1618** [comment]
It appears that support for Opus in an Ogg container was added specifically in Android 7.0:

- [CDD for 6.0](https://source.android.com/compatibility/6.0/android-6.0-cdd.html) only requires that Opus be supported in a Matroska container.
- [CDD for 7.0](https://source.android.com/compatibility/7.0/android-7.0-cdd.html) requires that Opus be supported in both Ogg and Matroska containers.

**Comment by timusus** [comment]
@bobobo1618 right, well that is good to know.

As far as this issue is concerned though - Shuttle supports Opus only as far as Android does, which is in a very limited fashion at the moment. Very savvy users might be able to encode their files using the appropriate container. 

The only solutions I know of to fully support Opus are:
- Android adds support natively
- Shuttle moves to a 3rd party playback engine.

**Comment by faho** [comment]
@Daykm: Can you try renaming your files to "*.ogg"? That worked for me, no reencoding necessary.

**Comment by bobobo1618** [comment]
> Very savvy users might be able to encode their files using the appropriate container.

As far as I'm aware, Ogg is the only appropriate container. It's the container used by `ffmpeg` and `opusenc` when the `.opus` file extension is selected as well as the one that [there's a spec for](https://wiki.xiph.org/OggOpus). There shouldn't be an encoder that outputs it as anything else (unless you're encoding video) so there shouldn't be any particularly hard work other than to make sure there's not a video stream (from embedded album art) and using the `.ogg` extension.

I actually did some more testing and it seems that Android's fine with there being a video stream in some circumstances. 

So the problems remaining as far as I'm aware are:
- Support Opus on Android < 7.0.
- Support Opus files with the `.opus` file extension.
- Support Opus files with album art erroneously transcoded as a video stream (the Ogg spec has provisions for album art as metadata), though I'd really treat this as as a problem with the encoder rather than the player. I'm still not certain what exactly causes it to fail in some cases.

The first certainly requires a 3rd party playback engine but I'm curious whether there's a way to hack around the second by writing your own scanner (and keeping Android's playback API) and the last by perhaps remuxing before sending to Android's playback API.

This is a lot of effort just to keep Android's playback API but I believe some devices, particularly on the lower end, have hardware audio decoders and I fear switching to software decode may negatively impact them.

**Comment by Hyleus** [comment]
I suggest to comment on or try to get these PRs merged: https://android-review.googlesource.com/q/owner:aykevanlaethem%2540gmail.com

If enough people make comments on the issue they will be forced sooner or later to merge it, they're ignoring it because they don't really care that much and most people don't even know what Opus is, hence they get currently away with it. It's such a small issue but still they ignored it for such a long time.
